### PR TITLE
Add cache Timeout configuration

### PR DESCRIPTION
**Pull request by martinpf**
This PR addresses the latest feedback from:
#603 
#608 

Fixes coding convention issues
Readds the console logging (logging will be addressed in another PR)
changes cacheTimeout  config item name to be more descriptive
Reinstates original cacheTimeout value

### CODE REVIEW

**Comment by SajayAntony** [comment]
@muratg Do you want to merge this into MusicStore dev?
@martinpf please squash all the commits and update the branch.

**Comment by muratg** [comment]
@kichalla Could you review and merge?

File: src/MusicStore/Startup.cs [hunk 1]
```diff
@@ -1,3 +1,4 @@
+using System.ComponentModel;
```
Comment (Eilon) [inline src/MusicStore/Startup.cs]: Why this namespace?

File: src/MusicStore/Properties/launchSettings.json [hunk 2]
```diff
@@ -17,6 +17,7 @@
     },
     "web": {
       "commandName": "web",
+      "commandLineArgs": " server.urls=http://*:5000  ",
```
Comment (Eilon) [inline src/MusicStore/Properties/launchSettings.json]: Trim spaces on this?

File: src/MusicStore/Startup.cs [hunk 3]
```diff
@@ -34,6 +35,7 @@ public Startup(IApplicationEnvironment applicationEnvironment, IRuntimeEnvironme
         public void ConfigureServices(IServiceCollection services)
         {
             services.Configure<AppSettings>(Configuration.GetSection("AppSettings"));
+            System.Console.WriteLine("Cache Timeout (Seconds):  " + Configuration["AppSettings:CacheTimeoutInSeconds"]);
```
Comment (Eilon) [inline src/MusicStore/Startup.cs]: Should use logging for this, not Console.WriteLine.

File: src/MusicStore/config.json [hunk 4]
```diff
@@ -1,14 +1,15 @@
 ﻿{
-    "AppSettings": {
-        "SiteTitle": "ASP.NET MVC Music Store"
-    },
+  "AppSettings": {
+    "SiteTitle": "ASP.NET MVC Music Store",
+    "CacheTimeoutInSeconds": "600"
```
Comment (Eilon) [inline src/MusicStore/config.json]: This is already the default value, no?

File: src/MusicStore/Areas/Admin/Controllers/StoreManagerController.cs [hunk 5]
```diff
@@ -8,8 +8,10 @@
 using Microsoft.AspNetCore.Mvc.Rendering;
 using Microsoft.EntityFrameworkCore;
 using Microsoft.Extensions.Caching.Memory;
+using Microsoft.Extensions.Options;
 using MusicStore.Models;
 using MusicStore.ViewModels;
+using MusicStore.Extensions;
```
Comment (Eilon) [inline src/MusicStore/Areas/Admin/Controllers/StoreManagerController.cs]: Sort & Remove namespaces in all updated files. (But keep System.* namespaces on top.)

File: src/MusicStore/Areas/Admin/Controllers/StoreManagerController.cs [hunk 6]
```diff
@@ -17,9 +19,12 @@ namespace MusicStore.Areas.Admin.Controllers
     [Authorize("ManageStore")]
     public class StoreManagerController : Controller
     {
-        public StoreManagerController(MusicStoreContext dbContext)
+        private AppSettings _appSettings;
```
Comment (Eilon) [inline src/MusicStore/Areas/Admin/Controllers/StoreManagerController.cs]: Can make this `readonly` (same in other controllers)

File: src/MusicStore/Controllers/HomeController.cs [hunk 7]
```diff
@@ -6,11 +6,22 @@
 using Microsoft.EntityFrameworkCore;
 using Microsoft.Extensions.Caching.Memory;
 using MusicStore.Models;
+using Microsoft.Extensions.Options;
+using MusicStore.Extensions;
+
```
Comment (Eilon) [inline src/MusicStore/Controllers/HomeController.cs]: Remove blank line.

File: src/MusicStore/Controllers/HomeController.cs [hunk 8]
```diff
@@ -6,11 +6,22 @@
...
 
 namespace MusicStore.Controllers
 {
     public class HomeController : Controller
     {
+
```
Comment (Eilon) [inline src/MusicStore/Controllers/HomeController.cs]: Remove blank line.

**Comment by Eilon** [comment]
:watch:

**Comment by SajayAntony** [comment]
ping @kichalla

**Comment by kichalla** [comment]
An alternative approach I am thinking about is to create a wrapper type for IMemoryCache where this type gets IMemoryCache and appsettings injected to it and does similar logic as the current extensions do. So where ever you want to use cache just get the wrapper type injected and that abstracts the details of checking the app setting etc.

File: src/MusicStore/config.json [hunk 9]
```diff
@@ -1,14 +1,16 @@
...
-        "DefaultConnection": {
-            // Use a shared (and running) LocalDB database when executing in IIS e.g.
-            // "Server=(localdb)\\.\\IIS_DB;Database=MusicStore;Trusted_Connection=False;MultipleActiveResultSets=true;User ID=iis_login;Password=********"
-            "ConnectionString": "Server=(localdb)\\MSSQLLocalDB;Database=MusicStore;Trusted_Connection=True;MultipleActiveResultSets=true;Connect Timeout=30;"
-        }
+      "DefaultConnection": {
+        // Use a shared (and running) LocalDB database when executing in IIS e.g.
+        // "Server=(localdb)\\.\\IIS_DB;Database=MusicStore;Trusted_Connection=False;MultipleActiveResultSets=true;User ID=iis_login;Password=********"
+        //"ConnectionString": "Server=(localdb)\\MSSQLLocalDB;Database=MusicStore;Trusted_Connection=True;MultipleActiveResultSets=true;Connect Timeout=30;",
+        "Connectionstring": "Server=martinpf-main2;Database=MusicStore;Trusted_Connection=True;MultipleActiveResultSets=true;Connect Timeout=30;"
```
Comment (SajayAntony) [inline src/MusicStore/config.json]: Why is this there?

File: src/MusicStore/Controllers/HomeController.cs [hunk 10]
```diff
@@ -28,12 +37,13 @@ public class HomeController : Controller
...
                     // Refresh it every 10 minutes.
                     // Let this be the last item to be removed by cache if cache GC kicks in.
-                    cache.Set(
+                    cache.SetExt(
                         cacheKey,
                         albums,
                         new MemoryCacheEntryOptions()
-                            .SetAbsoluteExpiration(TimeSpan.FromMinutes(10))
-                            .SetPriority(CacheItemPriority.High));
+                            .SetAbsoluteExpiration(TimeSpan.FromSeconds(_appSettings.CacheTimeoutSeconds > 0 ? _appSettings.CacheTimeoutSeconds : 1))
```
Comment (SajayAntony) [inline src/MusicStore/Controllers/HomeController.cs]: This seems odd. We should not be calling the expiration timeotu if there is none.

File: src/MusicStore/Extensions/MemoryCacheExtension.cs [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+
+namespace MusicStore.Extensions
+{
+    // Summary:
+    //     Specifying a cacheTimeout in config of 0 or less indicates the memory cache is not to be used at all.
+    //     Wrapper methods for the IMemoryCache service that stubs out the use of the cache if the cache timeout is 0 or less.
+    
+    public static class MemoryCacheExtensions
+    {
+        public static void RemoveExt(this IMemoryCache cache, object cacheKey, int cacheTimeout)
```
Comment (SajayAntony) [inline src/MusicStore/Extensions/MemoryCacheExtension.cs]: Why do we need the Ext postfix?

File: src/MusicStore/Properties/AppSettings.cs [hunk 12]
```diff
@@ -3,5 +3,6 @@
     public class AppSettings
     {
         public string SiteTitle { get; set; }
+        public int CacheTimeoutSeconds { get; set; } = 600;
```
Comment (kichalla) [inline src/MusicStore/Properties/AppSettings.cs]: If the idea is to just enable/disable cache, then probably requires a better name and bool value?

**Comment by Eilon** [comment]
@SajayAntony the change @kichalla would be proposes would be what we think is a cleaner way of doing this change.

**Comment by SajayAntony** [comment]
@Eilon closing this PR since the #620 provides the ability to disable the cache which is much simpler.

**Pull request closed without merging**